orttraining/orttraining/training_ops/cuda/activation/bias_gelu_grad_impl.cu

```diff
@@ -57,34 +57,45 @@ __global__ void BiasGeluGradDxKernel(int64_t bias_size, const T* dY, const T* X,
 template <typename T, typename GeluComputationMode>
 void LaunchBiasGeluGradDxKernel(
     cudaStream_t stream,
+    const cudaDeviceProp& prop,
     int64_t input_size, int64_t bias_size,
     const T* dY, const T* X, const T* B, T* dX) {
-  // given a 2D grid of blocks:
-  // each grid row handles bias_size elements
-  // there are input_size / bias_size rows
-  
+  // Given a logical 2D grid of blocks:
+  //   each grid row handles bias_size elements
+  //   there are input_size / bias_size rows
+  // For cases having many rows in the 2D grid (bigger than the upper limit of dimension of grid),
+  // 3D grid is used, e.g. the first two dimensions will be used as the logical "row".
+
+  const uint32_t maxGridY = static_cast<uint32_t>(prop.maxGridSize[1]);
   const int num_elements_per_thread = GridDim::maxElementsPerThread;
-  int max_threads_per_block = GridDim::maxThreadsPerBlock;
-  #ifdef USE_ROCM
+
+#ifdef USE_ROCM
   // Optimization for ROCm MI100
-  max_threads_per_block = 512;
-  #endif
-  
+  const int max_threads_per_block = 512;
+#else
+  const int max_threads_per_block = GridDim::maxThreadsPerBlock;
+#endif
+
   int num_threads_per_block =
-      std::min<int>(static_cast<int>(CeilDiv(bias_size, num_elements_per_thread)), static_cast<int>(max_threads_per_block));
+      std::min<int>(static_cast<int>(CeilDiv(bias_size, num_elements_per_thread)), max_threads_per_block);
   const auto grid_width = CeilDiv(bias_size, num_elements_per_thread * num_threads_per_block);
-  const auto grid_height = input_size / bias_size;
+  uint32_t grid_height = static_cast<uint32_t>(input_size / bias_size);
+  uint32_t grid_depth = 1;
+  if (grid_height >= maxGridY) {
+    grid_depth = static_cast<uint32_t>(CeilDiv(grid_height, maxGridY));
+    grid_height = maxGridY;
+  }
 
-  const dim3 grid_dim{static_cast<uint32_t>(grid_width), static_cast<uint32_t>(grid_height)};
+  const dim3 grid_dim{static_cast<uint32_t>(grid_width), grid_height, grid_depth};
```

Comment:
I am thinking if we could switch the width and height here in dim3 and modify the kernel to calculate the offset accordingly. max_grad_dim_x is a value big number, the (input_size/bias_size) is not likely to exceed that number. For max_grad_dim_y=65535, by considering the num_elements_per_thread=4 and num_threads_per_block=256, then the max bias_size is 67107840, which is also big enough, tensor is not likely to have such number. I am thinking this switching is more natual for understanding.

Reply:
Fair. Changed as suggested.